Add support for opensuse

mbgg opened the pull request:
Know issues:

1. adhd compilation with gcc7 is broken. Patch to fix this is submitted. As a workaround we would need to pass --Wno-error=int-in-bool-context (this holds for tumbleweed, leap 42.2 should work without a problem).
2. no audio working, need investigation

normal instalation should work without problems:
sudo sh crouton -r tumbleweed -n test -t xiwi,xfce,keyboard,cli-extra
sudo sh crouton -r leap-42.2 -n test -t xiwi,xfce,keyboard,cli-extra

dnschneid commented on installer/openSUSE/defaults:
```diff
@@ -0,0 +1,23 @@
+esac
+
+if [ -z "$MIRROR" ]; then
+    MIRROR="${CROUTON_MIRROR_opensuse_arm:-http://download.opensuse.org/ports/}"
```
Is this the right mirror for x86?

mbgg replied:
No, but actually MIRROR is not used anywhere. I'll delete it.

dnschneid commented on installer/openSUSE/defaults:
```diff
@@ -0,0 +1,23 @@
+
+case "$ARCH" in
+x86_64 | amd64) ARCH="x86_64";;
+arm*) ARCH="armv7l";;
```
Should this be armv7hl?

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh -e
+# Copyright (c) 2017 Matthias Brugger. All rights reserved.
```
Please release all code to be merged as copyright the crouton authors.

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+# Create release name from /etc/os-release file
+# sed specialist can make that more elegant...
+rel="`grep CPE_NAME "$sources" | cut -d ":" -f 4`"
+if [ $rel = "opensuse" ]
```
`if [ "$rel" = 'opensuse' ]; then`

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+    rel=tumbleweed
+elif [ $rel = "leap" ]
+then
+    ver=`grep CPE_NAME "$sources" | cut -d ":" -f 5 | sed 's/"$//'`
```
quote and `$( )`

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+fi
+
+# Create release name from /etc/os-release file
+# sed specialist can make that more elegant...
```
@drinkcat :)

DennisLfromGA replied:
Note to @drinkcat: add 'sed specialist' to [LinkedIn page](https://www.linkedin.com/in/nicolas-boichat-721a2622/). :)

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+    rel=${rel}-${ver}
+elif [ $rel = "tumbleweed" ]
+then
+    continue
```
`continue` doesn't work to exit an if statement.  If you want a no-op, use `:`, or make the exit 1 conditional on `rel != tumbleweed`.

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+# Why use sed if you know cut...
+sources="${2%/}/etc/products.d/openSUSE.prod"
+if [ "$1" = '-a' ]; then
+    grep "<arch>" "$sources"| cut -d '>' -f 2 | cut -d '<' -f 1
```
nit: space around `|`

dnschneid commented on installer/openSUSE/getrelease.sh:
```diff
@@ -0,0 +1,47 @@
+fi
+
+# Print the architecture if requested
+# Why use sed if you know cut...
```
what, you don't find it satisfying to spend an hour forcing a three-stage pipeline into a single command? ;)

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+# This is a distro-specific continuation of the prepare.sh script.
+
+PKGEXT='rpm'
+DISTROAKA='openSUSE'
```
let's keep non-user-visible references to openSUSE lowercased to avoid case-sensitivity issues.  That also includes the directory structure.

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+# install_dist: see install() in prepare.sh for details.
+install_dist() {
+    local pkgs='' pkgsasdeps='' params='' asdeps=''
+    while [ ! "$#" = 0 ]; do
```
`!=`...this is another thing we need to fix in the rest of the codebase :)

(here and elsewhere)

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+    fi
+    zypper -q install -y $params $pkgs `list_uninstalled_dist - "$@"` || return $?
+# TODO see how we can implement that reasonable with zypper
+# for now we just bloat the chroot
```
but at least you're honest about it!

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+        shift
+    fi
+    if [ ! "$#" = 0 ]; then
+#        rpm --nodeps -i "$@" || return $?
```
remove?

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+    then
+        echo "ERROR: Could not build dummy rpm!"
+    fi
+    pkg=$(awk '/^wrote:/ { print $2 }' < ${tmp}/build.log)
```
watch out for quotes

mbgg replied:
what do you mean exactly?

dnschneid replied:
`pkg="$(awk '/^wrote:/ { print $2 }' < "$tmp/build.log")"`

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+
+
+# list_uninstalled_dist: see list_uninstalled() in prepare.sh for details.
+# If the package is virtual (e.g. libc-dev), we need to find the binary package
```
does this comment apply?

dnschneid commented on installer/openSUSE/prepare:
```diff
@@ -0,0 +1,147 @@
+fi
+
+# enable network ping for all users
+sudo chmod u+s /usr/bin/ping
```
don't need sudo in chroot preparation; everything is already run as root

dnschneid commented on installer/openSUSE/releases:
```diff
@@ -0,0 +1 @@
+tumbleweed
```
what about `leap`?

mbgg replied:
it's not supported at the moment, I'll evaluate if dropping the leap part for now.

dnschneid commented on targets/audio:
```diff
@@ -99,6 +99,13 @@ build_cras() {
         ALSALIBDIR="/usr/lib$archextrapath/alsa-lib"
+    elif release -le tumbleweed; then
+        install --minimal --asdeps alsa-devel
+        if [ $ARCH = "aarch64" ] || [ $ARCH = "x86_64"]; then
```
`] || [` -> `-o`

Also, $ARCH needs quotes.

mbgg replied:
yes, I realized this tonight as well, most probably this is the reason why X is not working in targets/xiwi

dnschneid commented on targets/gtk-extra:
```diff
@@ -8,17 +8,19 @@ DESCRIPTION='GTK-based tools including gdebi, gksu, and a simple browser.'
-        break
-    fi
-done
+if release -ne tumbleweed; then
```
that will actually match all openSUSE releases that aren't tumbleweed.

Instead, you want `if ! release -eq tumbleweed`, or `if [ "$DISTROAKA" != opensuse ]` or something.

dnschneid replied:
Does opensuse not have an equivalent to alternatives?  How does it pick the most appropriate browser to use?

dnschneid commented on targets/post-common:
```diff
@@ -35,7 +35,12 @@ if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
 # Add the primary user
-groups='audio,input,video,wayland,sudo,plugdev,crouton'
+if [ "${DISTROAKA:-"$DISTRO"}" = 'openSUSE' ]; then
+    groups='audio,input,video,wayland,root,plugdev,crouton'
```
instead of two mostly-similar lists, why not put the common list above and just add sudo/root to groups in the if statements

mbgg replied:
sure

dnschneid commented on targets/x11-common:
```diff
@@ -7,6 +7,11 @@
 ### Append to prepare.sh:
+
+#we need to install xinit to be sure we overwrite the path
+if release -eq tumbleweed; then
```
anything that should happen on all opensuse releases should use the DISTROAKA check.

dnschneid commented on targets/xiwi:
```diff
@@ -38,7 +38,8 @@ fi
 # Compile croutonfbserver
 compile fbserver '-lX11 -lXfixes -lXdamage -lXext -lXtst' \
-    libx11-dev libxfixes-dev libxdamage-dev libxext-dev libxtst-dev
+    openSUSE=libX11-devel,libx11-dev openSUSE=libXfixes-devel,libsXxfixes-dev openSUSE=libXdamage-devel,libxdamage-dev \
```
`libsXxfixes-dev` is mangled

mbgg replied:
uups

dnschneid commented on targets/xiwi:
```diff
@@ -402,7 +403,13 @@ EOF
-if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+if [ "${DISTROAKA:-"$DISTRO"}" = 'openSUSE' ]; then
+    if [ $ARCH = "aarch64" ] || [ $ARCH = "x86_64"]; then
+        cp /usr/lib/xorg/modules/drivers/* /usr/lib64/xorg/modules/drivers/
```
Why is this necessary?

mbgg replied:
ok, I see, I should change the DRIVEDIR variable accordingly. I just wonder if on ubuntu/debian etc the drivers are installed under /usr/lib/... although the architecture is amd64. Can you confirm? Otherwise we could get rid of the first if, checking for openSUSE

mbgg replied:
It's for the Xorg dummy driver we install, all the other drivers should be fine.

dnschneid replied:
@drinkcat what's the best way to make sure the driver ends up in the right place?

dnschneid commented on targets/xiwi:
```diff
@@ -402,7 +403,13 @@ EOF
+    if [ $ARCH = "aarch64" ] || [ $ARCH = "x86_64"]; then
+        cp /usr/lib/xorg/modules/drivers/* /usr/lib64/xorg/modules/drivers/
+    else
+        cp /usr/lib/xorg/modules/drivers/* /usr/lib/xorg/modules/drivers/
```
this line doesn't make sense

mbgg replied:
ah, yes you are right :)

dnschneid commented on targets/xorg:
```diff
@@ -43,7 +43,7 @@ inteldriver=''
 pinmesa=''
 intelfbcsupport='y'
-if [ "${ARCH#arm}" = "$ARCH" ]; then
+if [ "${ARCH#arm}" = "$ARCH" ] && [ "${ARCH#aarch64}" = "$ARCH" ]; then
```
oh, this looks like a bug...could you make a separate PR for this?  I wonder if we've done similar mistakes elsewhere.

@drinkcat

dnschneid replied:
...or is it just that debian/ubuntu refers to it as arm64?

mbgg replied:
Well in install/debian/defaults we set ARCH to arm64 while I did set it to aarch64. We should unify this. I tend to use aarch64, as arm64 is only a kernel thing, but I don't care too much honestly.

dnschneid replied:
let's be consistent with the other distros and use arm64

mbgg replied:
ok, I'll fix installer/opensuse/getrelease.sh

dnschneid commented on targets/xorg:
```diff
@@ -68,7 +68,7 @@ if [ "${ARCH#arm}" = "$ARCH" ]; then
 
 # Catalog relevant and irrelevant video drivers
-fbdev="xserver-xorg-video-fbdev$backport"
+fbdev="openSUSE=xf86-video-fbdev,xserver-xorg-video-fbdev$backport"
```
I wonder if it would be cleaner to have each distro set the xorg package name prefix at the top and then just reference `$xorg-video-fbdev` (etc) without needing any distro modifiers

mbgg replied:
I don't see any schema that would be easy to implement. Some have a prefix, some have a postfix and some have a different name.

dnschneid commented on targets/xorg:
```diff
@@ -97,7 +97,7 @@ fi
 # On Freon, we need crazy xorg hacks
 if [ -n "$freon" ]; then
-    compile freon '-ldl -ldrm -I/usr/include/libdrm' so libdrm-dev
+    compile freon '-ldl -ldrm -I/usr/include/libdrm' so openSUSE=libdrm-devel,libdrm-dev
```
same goes for devel vs. dev, I guess.  You don't have to make any changes, though.

dnschneid commented on targets/xorg:
```diff
@@ -181,7 +181,7 @@ if [ -n "$backport" ]; then
 fi
 
-if release -lt stretch -lt kali-rolling -lt yakkety; then
+if release -lt stretch -lt kali-rolling -lt yakkety -lt tumbleweed; then
```
`-lt tumbleweed` means any opensuse release that is earlier than tumbleweed, not including tumbleweed itself.  Is that what you meant?

dnschneid replied:
note that if you don't want this for any opensuse release, you can just leave the line unmodified (without `-lt tumbleweed`) and it won't match opensuse.

dnschneid commented on targets/xorg:
```diff
@@ -194,12 +194,12 @@ if [ -z "$inteldriver" -a -n "$backport" ] && release -eq precise; then
     install_dummy xorg -- xserver-xorg$backport libgl1-mesa-glx$backport \
         libgl1-mesa-dri$backport libglu1-mesa xfonts-base x11-apps \
-        x11-session-utils x11-utils x11-xfs-utils x11-xkb-utils \
+        x11-session-utils openSUSE=xprop,x11-utils x11-xfs-utils x11-xkb-utils \
```
is xprop the only thing we need from x11-utils?

mbgg replied:
yes, AFAIK it is the only tool needed

dnschneid reviewed (changes requested):
This is a really great start; thanks!  Maybe this will beat gentoo and arch :)

The O of croagh!

mbgg commented on installer/openSUSE/debootstrap/debootstrap.sh:
```diff
@@ -0,0 +1,122 @@
+#
+# You should have received a copy of the GNU General Public License
+# A copy of the GNU General Public License can be downloaded from
+# http://www.gnu.org/licenses/gpl.html
```
Ok, I'll see how to resolve this.

mbgg commented:
Just a friendly ping that I send a new pull request in here (not sure if I screwed up because we lost the comments on the old one). So any comments are welcome :)

dnschneid commented on installer/opensuse/debootstrap.sh:
```diff
@@ -0,0 +1,76 @@
+#!/bin/sh -e
+# Copyright (c) 2017 The crouton Authors. All rights reserved.
```
You can't claim copyright just by deleting code, and most (all?) of this file appears to be copied from [here](https://github.com/openSUSE/debootstrap/blob/master/debootstrap.sh).

There's also no reason to fake the interface of debootstrap.  Please make an effort to implement opensuse/bootstrap without the use or reference of this code.  I suspect you can write what you need from scratch in a few lines inside opensuse/bootstrap.

mbgg committed:
installer: Add openSUSE distribution

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: post-common: openSUSE needs different groups

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: core: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: cli-extra: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: x11-common: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: audio: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: xorg: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: extension: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: gtk-extra: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: xiwi: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: xfce: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
targets: gnome: Enable openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
host-bin: enter-chroot: Enable systemd for openSUSE

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
chromium: Print distribution in error message

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg committed:
chromium: openSUSE does not support chromium

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

mbgg commented:
I updated the pull request. I dropped debootstrap and implemented the necessary bits directly in bootstrap. Also I added two commits for chromium (un)support.

Happly awaiting any comments :)

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+# from $RELEASE), $ARCH, and $MIRROR.
+
+# Tumbleweed
+url_tumbleweed_armv7hl="http://download.opensuse.org/ports/armv7hl/tumbleweed/images/openSUSE-Tumbleweed-ARM-JeOS.armv7hl-rootfs.armv7l-Current.tbz"
```
how did you decide on these URLs?  The sizes are dramatically different (the X11 ones are 800 megs!).  There're also distribution/openSUSE-current/appliances/ and distribution/openSUSE-stable/appliances/ directories with similar images.

There are unversioned entries in the leap subdirectory.  Are those always updated?  Are we going to have to keep updating the tarball in this file if we specify a version?

And what happens on i386/x86_64?

mbgg replied:
JeOS images are minimal images (Just Enough Operating System). For armv7hl no plain JeOS image exists. I'm investigating this. Maybe we can just delete the armv7hl support for Leap42.2 and add it in a later patch.

Leap images don't get updated, they get released and that's it. The unversioned images are links to the versioned ones. I'll update the links, as the version is of no use.

We should try to add a automatic resolution of architecutre + os-version -> download link, but I would prefer to leave this for a latter patch, e.g. when adding Leap-42.3 support. TBH I'm not too confident that this will work out, as the folder structure of the download server might change for newer releases.

i386/x86_64 images don't exists at this point in time, only the installer ISO is created.

mbgg replied:
so as we can't support i386 right now, I will delete the architecutre from installer/opensuse/defaults as well.

mbgg replied:
Ok, armv7hl image got created now, I will fix the URL.

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+BOOTSTRAP_RELEASE="$RELEASE"
+
+# Rename factory to tumbleweed
+[ "$RELEASE" = "factory" ] && BOOTSTRAP_RELEASE=tumbleweed
```
please convert this to a full if statement

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+[ "$ARCH" = "arm64" ] && BOOTSTRAP_ARCH=aarch64
+
+key=$(echo "\$url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}")
+URL=$(echo $(eval echo "${key}"))
```
I see what you're trying to do here, but considering all the setup and checking necessary, is this really shorter than just using if statements to explicitly set the URL?

dnschneid replied:
A new if statement (or a case statement) would probably be easier to follow, but maybe there's some complexity in that approach I'm not seeing.

mbgg replied:
I think with this approach, adding a new version/architecture is easier, as it is a one-line add.
When you put that in if statements, it's more complicated, as you have to figure out between different if's, the arch and distribution name.

dnschneid replied:
How about building up the URL in pieces?

mbgg replied:
Ok, I hacked that up for you, let me know if you prefer this, I definitely do not. File path and file names are not homogenous on architectures and releases. I expect this to break with a new architecture. Apart I'm quite sure that if you don't have the actual links in front of you, it is really hard to understand how this links will look like:

```bash
BOOTSTRAP_ARCH="$ARCH"
BOOTSTRAP_RELEASE="$RELEASE"
CURRENT=""

if [ "$ARCH" = "arm64" ] || [ "$ARCH" = "aarch64" ]; then
    BOOTSTRAP_ARCH="aarch64"
    BOOTSTRAP_ARCH2="aarch64"
    BOOTSTRAP_ARCH3="aarch64"
elif [ "$ARCH" = "armv7hl" ]; then
    BOOTSTRAP_ARCH="armv7hl"
    BOOTSTRAP_ARCH2="armv7hl"
    BOOTSTRAP_ARCH3="armv7l"
fi

if [ "$RELEASE" = "factory" ] || [ "$RELEASE" = "tumbleweed" ]; then
    BOOTSTRAP_VERSION="."
    BOOTSTRAP_RELEASE="tumbleweed/images"
    BOOTSTRAP_RELEASE2="."
    BOOTSTRAP_RELEASE3="Tumbleweed"
    CURRENT="-Current"
fi

case "$RELEASE" in leap*)
    BOOTSTRAP_VERSION=${RELEASE#leap-}
    BOOTSTRAP_RELEASE="distribution/leap"
    BOOTSTRAP_RELEASE2="appliances"
    BOOTSTRAP_RELEASE3="Leap$BOOTSTRAP_VERSION"

    if [ "$ARCH" = "armv7hl" ]; then
        BOOTSTRAP_ARCH2="armv7"
    fi
esac

URL="http://download.opensuse.org/ports/$BOOTSTRAP_ARCH/$BOOTSTRAP_RELEASE/$BOOTSTRAP_VERSION/$BOOTSTRAP_RELEASE2/openSUSE-$BOOTSTRAP_RELEASE3-ARM-JeOS.$BOOTSTRAP_ARCH2-rootfs.$BOOTSTRAP_ARCH3$CURRENT.tbz"
```

So for the record, here are the links we are talking about up to now:
http://download.opensuse.org/ports/armv7hl/tumbleweed/images/openSUSE-Tumbleweed-ARM-JeOS.armv7hl-rootfs.armv7l-Current.tbz
http://download.opensuse.org/ports/aarch64/tumbleweed/images/openSUSE-Tumbleweed-ARM-JeOS.aarch64-rootfs.aarch64-Current.tbz

http://download.opensuse.org/ports/armv7hl/distribution/leap/42.2/appliances/openSUSE-Leap42.2-ARM-JeOS.armv7-rootfs.armv7l.tbz
http://download.opensuse.org/ports/aarch64/distribution/leap/42.2/appliances/openSUSE-Leap42.2-ARM-JeOS.aarch64-rootfs.aarch64.tbz
http://download.opensuse.org/ports/aarch64/distribution/leap/42.3/appliances/openSUSE-Leap42.3-ARM-JeOS.aarch64-rootfs.aarch64.tbz

dnschneid replied:
Haha alright, I'm convinced.  Apologies for making you go through that.

mbgg replied:
No problem, the important thing is, that we reached consens on this :)

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+    exit 1
+fi
+
+mkdir -p $tmp/$subdir
```
quoting

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+fi
+
+mkdir -p $tmp/$subdir
+( cd $tmp/$subdir; wget -O - $URL | bzip2 -d | tar x ) 1>&2
```
quoting.  Also, `bzip2 -d | tar x` could just be `tar -jx`

dnschneid replied:
wget is being removed from the rootfs, so use curl:

`curl -# -L --connect-timeout 60 --retry 2 "$URL" | tar -jx -C "$tmp/$subdir"`

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+mkdir -p $tmp/$subdir
+( cd $tmp/$subdir; wget -O - $URL | bzip2 -d | tar x ) 1>&2
+
+echo 'Remove /dev files as not needed'
```
I don't think the user needs to see this

dnschneid commented on installer/opensuse/defaults:
```diff
@@ -0,0 +1,19 @@
+# It must set at least ARCH and MIRROR if not already specified.
+
+if [ -z "$ARCH" ]; then
+    ARCH="`uname -m`"
```
use `$( )`

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+URL=$(echo $(eval echo "${key}"))
+
+if [ ! "$URL" ]; then
+    echo "Unknown Distribution / Architecture: $key"
```
this should really be failing in `defaults`, not in bootstrap

dnschneid commented on installer/opensuse/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+
+# Create release name from /etc/os-release file
+# sed specialist can make that more elegant...
+rel=$(grep CPE_NAME "$sources" | cut -d ":" -f 4)
```
quoting

dnschneid commented on installer/opensuse/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+# Create release name from /etc/os-release file
+# sed specialist can make that more elegant...
+rel=$(grep CPE_NAME "$sources" | cut -d ":" -f 4)
+if [ $rel = "opensuse" ]; then
```
quoting...

dnschneid replied:
going to stop pointing it out here...but you need to make sure all your variable expansions are properly quoted

mbgg replied:
ok, hopefully I will catch all of them.

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+        pkgsasdeps="`list_uninstalled_dist '' $pkgs`"
+    fi
+    zypper -q install -y $params $pkgs `list_uninstalled_dist - "$@"` || return $?
+# TODO see how we can implement that reasonable with zypper
```
what's the plan here?

dnschneid replied:
Sure; the resulting chroot will just be larger because of it.

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+        fi
+        shift
+    done
+    while [ ! "$#" = 0 ]; do
```
!=

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+        params='--no-recommends'
+        shift
+    fi
+    if [ ! "$#" = 0 ]; then
```
!=

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+
+%files
+EOF
+    rpmbuild --define='_rpmdir ${tmp}' -bb ${tmp}/dummy.spec > ${tmp}/build.log
```
quoting.  also, is single quotes correct for --define?

mbgg replied:
yes, man page shows single quotes as well.

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+%files
+EOF
+    rpmbuild --define='_rpmdir ${tmp}' -bb ${tmp}/dummy.spec > ${tmp}/build.log
+    if [ $? != 0 ]
```
if [ ]; then

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+    done
+}
+
+if [ -r /debootstrap.sh ]; then
```
not needed?

mbgg replied:
no, not needed.

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+chmod u+s /usr/bin/ping
+
+# add rpmbuild for dummy packages
+sudo zypper -q install -y --no-recommends rpmbuild
```
no need for sudo.  also, shouldn't this be an `install rpmbuild` command in core?

dnschneid commented on targets/chromium:
```diff
@@ -3,9 +3,9 @@
+        [ "$DISTRO" = 'debian' -o "$DISTRO" = 'kali' -o "$DISTRO" = 'opensuse' ] &&
         [ "${ARCH#arm}" != "$ARCH" ]; then
-    error 99 "chromium target is not supported on Debian/ARM."
+    error 99 "chromium target is not supported on "$DISTRO"/"$ARCH"."
```
this quoting isn't quite what you want

mbgg replied:
It's only available for x86, not for ARM.

mbgg replied:
would something like this be acceptable for you:
    error 99 "chromium target is not supported on "$DISTRO"/ARM."

dnschneid replied:
you don't need the quotes around $DISTRO, since the entire string is quoted

dnschneid replied:
and I'd prefer "chromium target is not supported on $DISTRO on ARM"

mbgg replied:
sure, no problem

dnschneid commented on targets/chromium:
```diff
@@ -3,9 +3,9 @@
+        [ "$DISTRO" = 'debian' -o "$DISTRO" = 'kali' -o "$DISTRO" = 'opensuse' ] &&
         [ "${ARCH#arm}" != "$ARCH" ]; then
-    error 99 "chromium target is not supported on Debian/ARM."
+    error 99 "chromium target is not supported on "$DISTRO"/"$ARCH"."
```
is it really not available though?  seems like it exists, although I have no idea what it involves: https://software.opensuse.org/package/chromium

dnschneid commented on targets/x11-common:
```diff
@@ -7,6 +7,12 @@
 
 ### Append to prepare.sh:
+
+#we need to install xinit to be sure we overwrite the path
```
what does this mean

mbgg replied:
TBH I can't remember, but we need to install xinit, otherwise graphic does not work.
Do you prefer that I add this line instead of the if:
install opensuse=xinit,

Just let me know.

dnschneid replied:
I'd prefer to not install things unnecessarily, so please check what xinit is pulling in when you try to install it here

mbgg replied:
It's not pulling in anything, no recommends and sugguests:

linux@localhost:~> zypper info --recommends --suggests xinit
Loading repository data...
Reading installed packages...


Information for package xinit:
------------------------------
Repository     : openSUSE-Ports-Leap-42.3-repo-oss
Name           : xinit                            
Version        : 1.3.4-3.1                        
Arch           : aarch64                          
Vendor         : openSUSE                         
Installed Size : 234.6 KiB                        
Installed      : Yes (automatically)              
Status         : up-to-date                       
Source package : xinit-1.3.4-3.1.src              
Summary        : X Window System initializer      
Description    :                                  
    The xinit program is used to start the X Window System server and a
    first client program on systems that are not using a display manager
    such as xdm or in environments that use multiple window systems.
    When this first client exits, xinit will kill the X server and then
    terminate.
Recommends     : ---                              
Suggests       : ---

dnschneid replied:
Ah, I see.  This is fine, then.  Just move it under the "remove old vgem hack" section and quote $DISTROAKA

dnschneid replied:
Oh weird, we're already installing xinit in the xorg.  Is this an issue for xorg or only for xiwi?

dnschneid replied:
...in which case, xiwi line 31 is probably the right place for it.

mbgg replied:
ok, I put it at line 31 in xiwi.

mbgg commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+[ "$ARCH" = "arm64" ] && BOOTSTRAP_ARCH=aarch64
+
+key=$(echo "\$url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}")
+URL=$(echo $(eval echo "${key}"))
```
The idea behind that is, that if we can't be sure the download links keep the same folder structure over time, we just add a new identifier for, let's say leap-42.3 and we are fine.

This means that we need to convert "leap-42.3" to "leap_42_3" which is done in a dump way up to now. For the next revision I will change this using sed. So in effect if a new Leap release is added, we just need to put url_leap_42_3_aarch64 URL in the file and we are settled. This is still worse then other distributions in crouton, but IMHO better then start adding if statements of every new release. Thoughts?

mbgg commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,142 @@
+        pkgsasdeps="`list_uninstalled_dist '' $pkgs`"
+    fi
+    zypper -q install -y $params $pkgs `list_uninstalled_dist - "$@"` || return $?
+# TODO see how we can implement that reasonable with zypper
```
There is none such functionality in zypper, I think we should get rid of all the asdeps stuff in this file. Thoughts?

mbgg committed:
installer: opensuse: add Leap 42.3

Signed-off-by: Matthias Brugger <mbrugger@suse.com>

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+# fix up the good old toolchain/kernel naming conflict
+[ "$ARCH" = "arm64" ] && BOOTSTRAP_ARCH=aarch64
+
+key=$(echo "\$url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}")
```
I *think* these two lines can be replaced with:

`eval "URL=\"\$url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}\""`

dnschneid replied:
or to handle the nonexistent case better:

`eval "URL=\"\${url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}-}\""`

which just sets URL to an empty string.

mbgg replied:
yes, will do so.

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,41 @@
+key=$(echo "\$url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}")
+URL=$(echo $(eval echo "${key}"))
+
+if [ ! "$URL" ]; then
```
-z

mbgg replied:
done

mbgg commented:
@dnschneid , I think there was an overlap between you adding more comments on the code and me updating the pull request. I integrated your last two comments. Please feel free to review :)

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+fi
+
+# Fix leap release name
+case "$RELEASE" in leap*)
```
elif [ "${RELEASE#leap}" != "$RELEASE" ]; then

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+
+# Fix leap release name
+case "$RELEASE" in leap*)
+     BOOTSTRAP_RELEASE=$(echo "$RELEASE" | sed -e 's/[-.]/_/g')
```
quoting, and tr would probably be a better choice

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+eval "URL=\"\${url_${BOOTSTRAP_RELEASE}_${BOOTSTRAP_ARCH}-}\""
+
+if [ -z "$URL" ]; then
+    echo "Unknown Distribution / Architecture: $key"
```
if the `error` function is available, you probably want to use that.  Otherwise, you want to redirect to STDERR using `>&2`

Also, `$key` doesn't exist anymore

mbgg replied:
it seems, that no error function is available.

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+    exit 1
+fi
+
+mkdir -p "$tmp"/"$subdir"
```
`"$tmp/$subdir"`

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+fi
+
+mkdir -p "$tmp"/"$subdir"
+( cd "$tmp"/"$subdir"; wget -O - $URL |  tar -jx ) 1>&2
```
wget is being removed from the rootfs, so use curl:

`curl -# -L --connect-timeout 60 --retry 2 "$URL" | tar -jx -C "$tmp/$subdir"`

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,48 @@
+mkdir -p "$tmp"/"$subdir"
+( cd "$tmp"/"$subdir"; wget -O - $URL |  tar -jx ) 1>&2
+
+rm -rf "$tmp"/"$subdir"/dev/*
```
`"$tmp/$subdir/dev/"*`

dnschneid commented on installer/opensuse/defaults:
```diff
@@ -0,0 +1,18 @@
+
+case "$ARCH" in
+arm*) ARCH="armv7hl";;
+arm64 | aarch64) ARCH="arm64";;
```
I think this needs to be above the arm* line, otherwise arm64 won't be matched properly

mbgg replied:
good catch!

dnschneid commented on installer/opensuse/defaults:
```diff
@@ -0,0 +1,18 @@
+case "$ARCH" in
+arm*) ARCH="armv7hl";;
+arm64 | aarch64) ARCH="arm64";;
+*) error 2 "Invalid architecture '$ARCH'.";;
```
wait what happened to the x86 architectures

mbgg replied:
I dropped it for now, as there does no JeOS image exist, but just a installer image. That's something for the future. Creating an x86 image in the open build service and use it.

dnschneid replied:
definitely need x86 support before we can merge; this would confuse most of our users

dnschneid commented on installer/opensuse/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+    rel=tumbleweed
+elif [ "$rel" = "leap" ]; then
+    ver="$(grep CPE_NAME "$sources" | cut -d ":" -f 5 | sed 's/"$//')"
+    rel="$rel"-"$ver"
```
```
ver="$(grep CPE_NAME "$sources" | cut -d ":" -f 5)"
rel="$rel-${ver%\"}"
```

dnschneid commented on installer/opensuse/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+elif [ "$rel" = "leap" ]; then
+    ver="$(grep CPE_NAME "$sources" | cut -d ":" -f 5 | sed 's/"$//')"
+    rel="$rel"-"$ver"
+elif [ "$rel" != "tumbleweed" ]; then
```
make this "else" and add `-o "$rel" = 'tumbleweed'` to the first block

dnschneid commented on installer/opensuse/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+
+# Print the architecture if requested
+# Why use sed if you know cut...
+sources="${2%/}/etc/products.d/openSUSE.prod"
```
fail if this doesn't exist?

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+            params='--no-recommends'
+        elif [ "$1" = "--asdeps" ]; then
+            shift
+	    continue
```
indentation

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+            break
+        fi
+        pkgs="$pkgs $1"
+       shift
```
indentation

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+        pkgs="$pkgs $1"
+       shift
+    done
+    zypper -q install -y $params $pkgs `list_uninstalled_dist - "$@"` || return $?
```
use `$( )` instead of `` ` ` ``

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+        done
+    fi
+
+    if [ $pkgdepends ]; then
```
`if [ -n "$pkgdepends" ]; then`

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+        requires="Requires: ${pkgdepends}"
+    fi
+
+    local tmp="`mktemp -d crouton.XXXXXX --tmpdir=/tmp`"
```
use `$( )`

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+Provides: $pkgprovides
+$requires
+Summary: Provides a dummy ${pkgname#*-} for crouton
+BuildRoot:      %{_tmppath}/%{name}-%{version}-build
```
extra spaces unnecessary?

mbgg replied:
no

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+
+%files
+EOF
+    rpmbuild --define='_rpmdir "$tmp"' -bb "$tmp"/dummy.spec > "$tmp"/build.log
```
quotes should surround the whole string (e.g., `"$tmp/dummy.spec"`) as long as wildcards aren't involved

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+EOF
+    rpmbuild --define='_rpmdir "$tmp"' -bb "$tmp"/dummy.spec > "$tmp"/build.log
+    if [ $? != 0 ]; then
+        echo "ERROR: Could not build dummy rpm!"
```
non-fatal?

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,129 @@
+        echo "ERROR: Could not build dummy rpm!"
+    fi
+    pkg="$(awk '/^wrote:/ { print $2 }' < "$tmp/build.log")"
+    rm "$tmp"/build.log
```
quote surround string

dnschneid commented on installer/opensuse/releases:
```diff
@@ -0,0 +1,3 @@
+tumbleweed
+leap-42.2
+leap-42.3
```
and we don't want just a "leap" release that grabs the latest version, right?

mbgg replied:
no, these are independent releases.

dnschneid commented on targets/audio:
```diff
@@ -99,6 +99,13 @@ build_cras() {
         ALSALIBDIR="/usr/lib$archextrapath/alsa-lib"
+    elif [ "$DISTROAKA" = "opensuse" ]; then
+        install --minimal --asdeps alsa-devel
+        if [ "$ARCH" = "arm64" ] || [ "$ARCH" = "x86_64"]; then
```
`] || [` should be `-o`

dnschneid commented on targets/chromium:
```diff
@@ -3,9 +3,9 @@
 # found in the LICENSE file.
 if [ "${TARGETNOINSTALL:-c}" = 'c' ] &&
-        [ "$DISTRO" = 'debian' -o "$DISTRO" = 'kali' ] &&
+        [ "$DISTRO" = 'debian' -o "$DISTRO" = 'kali' -o "$DISTRO" = 'opensuse' ] &&
```
not distroaka?

dnschneid commented on targets/gtk-extra:
```diff
@@ -8,17 +8,19 @@ DESCRIPTION='GTK-based tools including gdebi, gksu, and a simple browser.'
 ### Append to prepare.sh:
 install_dummy network-manager network-manager-gnome
-install gdebi gksu
+    install opensuse=,gdebi opensuse=libgnomesu,gksu
```
indentation

dnschneid commented on targets/xiwi:
```diff
@@ -29,6 +29,13 @@ if [ "${DISTROAKA:-"$DISTRO"}" = 'debian' ]; then
+
+# We need to install xinit to make graphics work
+if [ $DISTROAKA = "opensuse" ]; then
+    install xinit
```
Just add it as `opensuse=xinit,` in the install line below

dnschneid commented on targets/xorg:
```diff
@@ -194,11 +196,11 @@ if [ -z "$inteldriver" -a -n "$backport" ] && release -eq precise; then
         xterm x11-common xinput
 else
-    install xorg $installvideodrivers -- xserver-xorg-video-all$backport \
+    install opensuse=xorg-x11-server,xorg $installvideodrivers -- opensuse=xorg-x11-driver-video,xserver-xorg-video-all$backport \
```
probably need to break the line

dnschneid reviewed (changes requested):
Lots of nits and a few minor errors, but no major concerns!  Woo!

mbgg commented:
Hi @dnschneid just now the newest version. Feel free to comment :)

mbgg commented:
Gentle ping, would be nice if you find a moment to review :-)

mbgg commented:
any further comments on that?

dnschneid commented on installer/opensuse/bootstrap:
```diff
@@ -0,0 +1,46 @@
+if [ "$RELEASE" = "factory" ]; then
+    BOOTSTRAP_RELEASE=tumbleweed
+elif [ "${RELEASE#leap}" != "$RELEASE" ]; then
+     BOOTSTRAP_RELEASE=$(echo "$RELEASE" | tr - _ | tr . _)
```
quoting (and line below)

dnschneid commented on installer/opensuse/prepare:
```diff
@@ -0,0 +1,130 @@
+    rpmbuild --define='_rpmdir "$tmp"' -bb "$tmp/dummy.spec" > "$tmp/build.log"
+    if [ $? != 0 ]; then
+        echo "ERROR: Could not build dummy rpm!"
+	exit 1
```
spaces, not tabs

dnschneid reviewed (changes requested):
Sorry I'm so slow at looking at this, especially since you've done such a great job cleaning it up.

It would be helpful if you avoid squashing fixes back into old commits (maybe until it's approved for merge and then you can rebase and make it pretty), since that causes the Github review UI to drop most of the history and comments and results in me needing to do a full review every time you update.
